### PR TITLE
Allow reference types for pinned GC.AllocateArray()

### PR DESCRIPTION
**Pull request by jthorborg**
After an offline discussion, this is the "... or similar" part of the suggested feature set in this issue: https://github.com/dotnet/runtime/discussions/48703

This just relaxes front-end constraints as agreed for pinned array allocations with ref types, which is already possible and utilized internally.

The use case shortly is forming unmanaged references to array indices, where the array has a very long lifetime.

I added some extra coverage of the API, noticing the original constraint wasn't tested in the `System.Runtime` tests. I'm not sure if further documentation edits are necessary, or if it's fine to just remove the `<remarks>` section.

### CODE REVIEW

**Comment by jthorborg** [comment]
@dotnet-policy-service agree company="Unity Technologies"

**Comment by jkotas** [comment]
The same fix should be also applied to:
src\coreclr\nativeaot\System.Private.CoreLib\src\System\GC.NativeAot.cs
src\mono\System.Private.CoreLib\src\System\GC.Mono.cs

**Comment by jthorborg** [comment]
> The same fix should be also applied to: src\coreclr\nativeaot\System.Private.CoreLib\src\System\GC.NativeAot.cs src\mono\System.Private.CoreLib\src\System\GC.Mono.cs

Done, thanks I missed that. Are there tests to update for these as well?

**Comment by MichalPetryka** [comment]
I think that AllocateUninitializedArray should also be relaxed to allow this since it already accepts reference types.

**Comment by jthorborg** [comment]
I considered that but I worried that garbage bits on reference fields would perhaps not be good news for the GC.

**Comment by EgorBo** [comment]
Seems like there is a test that expects that exception - `AllocateArrayUninitializedPinned_RefType_ThrowsArgumentException`

> I think that AllocateUninitializedArray should also be relaxed to allow this since it already accepts reference types.

Sounds like a very bad idea to me

**Comment by MichalPetryka** [comment]
> (unless we want it to work but it will never return garbage)

That's how it behaves today:
https://github.com/dotnet/runtime/blob/79823768e461a865019d7bdf70cfe0fcbeb99288/src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs#L741-L746

> I considered that but I worried that garbage bits on reference fields would perhaps not be good news for the GC.

Seeing above, I'd say that making the pinned path forward to `AllocateArray` and return cleared memory makes sense here.

**Comment by jthorborg** [comment]
Okay, I've similarly relaxed `AllocateUninitializedArray` as well by delegating to `AllocateArray` if it's a reference type. I think it's a reasonable affordance for users, given it's not a promise to return uninitialized memory anyway from this API.

I refactored this tests and combined them to parametric ones to take this into account.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 1]
```diff
@@ -759,7 +756,7 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
             }
             else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
+                return AllocateArray<T>(length, pinned: true);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:759]:
```suggestion
                // Deferring to the initializing allocation API since elements
                // may contain references.
                return AllocateArray<T>(length, pinned: true);
```

If the above isn't appropriate, then I think it would make sense to just set `pinned = true;` and fall through.

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:759]: The whole `RuntimeHelpers.IsReferenceOrContainsReferences<T>()` path can be deleted. 

The VM is going to force zero-initialization for arrays with references here: https://github.com/dotnet/runtime/blob/a864ec70925181f00c38e15e4b1896f2584953c4/src/coreclr/vm/gchelpers.cpp#L217-L218

**Comment by jkotas** [comment]
> Done, thanks I missed that. Are there tests to update for these as well?

The tests are shared for all runtime flavors. The test updates that you have done should be sufficient.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 1 above)
Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:759]: I considered that but not sure about what `GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL` means for ref types. I assumed since the frontend protects this path it may go bad.

**Comment by jthorborg** [comment]
> > Done, thanks I missed that. Are there tests to update for these as well?
> 
> The tests are shared for all runtime flavors. The test updates that you have done should be sufficient.

Very nice!

**Comment by cshung** [comment]
@Maoni0 - this change relaxes the constraint that user POH allocation can contain references. This might impact our plan for https://github.com/dotnet/runtime/pull/86802?

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 1 above)
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:759]: `GC_ALLOC_ZEROING_OPTIONAL` is effectively ignored for ref types. It should be fine to set it unconditionally here. The VM code that I have linked above should handle it.

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:759]: > I assumed since the frontend protects this path it may go bad.

The frontend check was there to give us freedom to drop the support in the GC implementation if we find it convenient.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 2]
```diff
@@ -732,9 +732,6 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
         /// <typeparam name="T">Specifies the type of the array element.</typeparam>
         /// <param name="length">Specifies the length of the array.</param>
         /// <param name="pinned">Specifies whether the allocated array must be pinned.</param>
-        /// <remarks>
-        /// If pinned is set to true, <typeparamref name="T"/> must not be a reference type or a type that contains object references.
-        /// </remarks>
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // forced to ensure no perf drop for small memory buffers (hot path)
         public static unsafe T[] AllocateUninitializedArray<T>(int length, bool pinned = false) // T[] rather than T?[] to match `new T[length]` behavior
         {
```
Comment (MichalPetryka) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:737]: Maybe remove the check in the `!pinned` path too?

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 1 above)
Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs]: Alright good to know. In that case, I simplified the paths in NativeAOT and the normal CoreCLR.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 2 above)
Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:737]: Let me know if 13fe7a16ae5407a1b8b4d91f93a14e6fa2d1714b is what you meant.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 3]
```diff
@@ -757,16 +754,10 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
             }
-            else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-            {
-                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
-            }
 
-            GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
-            if (pinned)
-                flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
+            GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL | GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:758]:
```suggestion
            GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
            
            if (pinned)	
                flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
```
We do not want to be pinning the array unconditionally. This should be kept intact.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs [hunk 4]
```diff
@@ -819,19 +816,13 @@ public static unsafe T[] AllocateUninitializedArray<T>(int length, bool pinned =
...
-            return AllocateNewUninitializedArray(length, pinned);
+            return AllocateNewUninitializedArray(length);
 
-            static T[] AllocateNewUninitializedArray(int length, bool pinned)
+            static T[] AllocateNewUninitializedArray(int length)
             {
-                GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
-                if (pinned)
-                    flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
+                GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL | GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs:825]: Same here.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 2 above)
Comment (MichalPetryka) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:737]: No, I've meant to remove the `RuntimeHelpers.IsReferenceOrContainsReferences<T>()` check there.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 3 above)
Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs]: Yeah, last commit was rushed. Please check latest (should fix below as well).

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 2 above)
Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:737]: I see what you mean. Check latest commit.

File: src/libraries/System.Runtime/tests/System/GCTests.cs [hunk 5]
```diff
@@ -1058,11 +1058,52 @@ private static void AllocateArrayTooLarge()
...
+            TryType<object>();
+        }
+
+        [Theory]
+        [InlineData(false), InlineData(true)]
+        private unsafe static void AllocateArrayPinned_ManagedValueType_CanRoundtripThroughPointer(bool uninitialized)
         {
-            GC.AllocateUninitializedArray<string>(100);
-            Assert.Throws<ArgumentException>(() => GC.AllocateUninitializedArray<string>(100, pinned: true));
+            const int k_Length = 100;
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1088]:
```suggestion
            const int Length = 100;
```
Nit: We do not use `k_` prefix for constants in this repo.

Comment (jthorborg) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1088]: Do you have a coding style reference? What should I change it to, just `length`?

Comment (jkotas) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1088]: Yes, `length` is fine.

**Comment by Maoni0** [comment]
@cshung I thought the new fix for that (after [#87533](https://github.com/dotnet/runtime/pull/87533)) is just to do the marking we are doing currently in `allocate_uoh_object` while holding the msl, am I missing something?

**Comment by cshung** [comment]
> @cshung I thought the new fix for that (after [#87533](https://github.com/dotnet/runtime/pull/87533)) is just to do the marking we are doing currently in `allocate_uoh_object` while holding the msl, am I missing something?

Nope, you are not. I just wanted to make sure you are aware of this, it was unclear you were in this thread. I assume the new plan (i.e. do the marking while holding msl) will work even when the user allocation could contain references?

**Comment by Maoni0** [comment]
> Nope, you are not. I just wanted to make sure you are aware of this, it was unclear you were in this thread.

actually I responded internally when you asked about it this morning. maybe you didn't notice the Teams message?

> I assume the new plan (i.e. do the marking while holding msl) will work even when the user allocation could contain references?

if you look at the code in `allocate_uoh_object`, it doesn't care if the object has references or not and objects on LOH can certainly have references. same principle works for POH objects.

File: src/libraries/System.Runtime/tests/System/GCTests.cs [hunk 6]
```diff
@@ -1058,11 +1058,52 @@ private static void AllocateArrayTooLarge()
             Assert.Throws<OutOfMemoryException>(() => GC.AllocateUninitializedArray<double>(int.MaxValue, pinned: true));
         }
 
-        [Fact]
-        private static void AllocateArrayRefType()
+        struct EmbeddedValueType<T>
+        {
+            unsafe fixed byte _[7];
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: Is there a specific reason for this? It looks like trying to fiddle with offsets and alignment. I tend to avoid these kind of mechanisms in tests without a comment as to the why.

**Comment by AaronRobinsonMSFT** [comment]
I've no concerns with this other than a question on the test. I defer to @Maoni0 and @jkotas for the final sign off.

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (jthorborg) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: It was a simple way of increasing the size of the struct, making it less trivially laid out and requiring padding bits - to add a bit of confidence in we're not just overwriting the first set of bytes in a memory location.

We could add a comment or just remove it, it's not required.

Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1063]:
```suggestion
            // Adding a simple field to force non-zero object offsets
            int _;
```

I assume the above is sufficient for these purposes.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 7]
```diff
@@ -757,16 +754,12 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
-            {
-                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
-            }
 
             GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
             if (pinned)
                 flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
 
-            return Unsafe.As<T[]>(AllocateNewArray(RuntimeTypeHandle.ToIntPtr(typeof(T[]).TypeHandle), length, flags));
+            return Unsafe.As<T[]>(AllocateNewArray(typeof(T[]).TypeHandle.Value, length, flags));
```
Comment (Maoni0) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:762]: nit - I would just add a comment here to say that the runtime will always clear the `GC_ALLOC_ZEROING_OPTIONAL` flag if a type contains refs, just to make it clear that GC will always memory for types containing refs (which always needs to be enforced) so you don't need to filter it here.

**Review by Maoni0 — APPROVED** [review]
other than my nit comment it LGTM! thanks for doing this work @jthorborg!

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (jthorborg) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: A single byte might fulfil the same purposes. Not sure about whether the VM will actually rearrange references fields to the start, actually. Or is that only for auto-layout (not default for structs)?

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 7 above)
Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:762]:
```suggestion
            // Runtime overrides GC_ALLOC_ZEROING_OPTIONAL if the type contains references, so we don't need to worry about that.
            return Unsafe.As<T[]>(AllocateNewArray(typeof(T[]).TypeHandle.Value, length, flags));
```
?

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (MichalPetryka) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > Or is that only for auto-layout (not default for structs)?

Managed structs always have Auto layout (or Explicit).

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > > Or is that only for auto-layout (not default for structs)?
> 
> Managed structs always have Auto layout (or Explicit).

That is not true. C# `struct`s or .NET ValueTypes, by default, are marked as `LayoutKind.Sequential`. C# `class`s or .NET types derived from `System.Object` are marked as `LayoutKind.Auto` by default.

Comment (MichalPetryka) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > are marked as `LayoutKind.Sequential`

Yes but the VM ignores this, see this sample:
```C#
[StructLayout(LayoutKind.Sequential)] struct S1 { public nint a; public object b; }
[StructLayout(LayoutKind.Sequential)] struct S2 { public nint a; public nint b; }
S1 s = new S1 { a = -1, b = null };
Console.WriteLine(Unsafe.As<S1, S2>(ref s).b);
```
It prints -1, which indicates that in S1 field b was placed first.
If you do:
```C#
[StructLayout(LayoutKind.Explicit)] struct S1 { [FieldOffset(0)] public nint a; [FieldOffset(8)] public object b; }
[StructLayout(LayoutKind.Explicit)] struct S2 { [FieldOffset(0)] public nint a; [FieldOffset(8)] public nint b; }
S1 s = new S1 { a = -1, b = null };
Console.WriteLine(Unsafe.As<S1, S2>(ref s).b);
```
You'll then get the correct 0 which indicates that Explicit layout is respected.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > Yes but the VM ignores this, see this sample:

The VM absolutely does not ignore this, it is incredibly important for type layouts and a core concept in interop. The example above is fundamentally flawed because the type is non-blittable. Non-blittable types are not normally involved in interop scenarios therefore things can change. This PR is making an assumption on that, which has been deemed fine and calculated. However, making that statement based on a single observation is flawed reasoning.

See https://github.com/dotnet/runtime/blob/531ad9596fef94194744ccfcac7159d1c1ae5e57/src/coreclr/vm/methodtablebuilder.cpp#L1760 and https://github.com/dotnet/runtime/blob/531ad9596fef94194744ccfcac7159d1c1ae5e57/src/coreclr/vm/methodtablebuilder.cpp#L11565.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > This PR is making an assumption on that, which has been deemed fine and calculated.

We actually might want to validate the layout scenarios however. If the type loader is altering layouts this PR becomes a bit more controversial.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: I should also apologize for being a overly aggressive on this one. I'm sorry. The statement out of context is what I was responding to and that is wrong.

If non-blittable valuetypes don't respect the sequential layout, then I don't fully understand how this sort of offset indexing will work reliably. It would appear we would be at the mercy of the typeloaders layout algorithm, which has been known to change. This actually might be worth considering a bug that needs to be fixed. I defer to the type system people though.

Comment (MichalPetryka) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > The example above is fundamentally flawed because the type is non-blittable.

Like I said originally, I was talking about "managed value types" which are by definition non-blittable.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > "managed value types" which are by definition non-blittable.

I've never heard of a "managed value type". Value types can have managed memory fields, but a value type can be on the stack (managed by the stack lifetime) or within an array (managed by the GC). Can you share where this term is used?

Comment (jkotas) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > managed value type

Managed vs. unmanaged types is C# terminology: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/unmanaged-types

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 8]
```diff
@@ -757,16 +754,13 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
-                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
-            }
 
             GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
             if (pinned)
                 flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
 
-            return Unsafe.As<T[]>(AllocateNewArray(RuntimeTypeHandle.ToIntPtr(typeof(T[]).TypeHandle), length, flags));
+            // Runtime overrides GC_ALLOC_ZEROING_OPTIONAL if the type contains references, so we don't need to worry about that.
+            return Unsafe.As<T[]>(AllocateNewArray(typeof(T[]).TypeHandle.Value, length, flags));
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:763]: Why did you change `RuntimeTypeHandle.ToIntPtr` to `RuntimeTypeHandle.Value`?

`RuntimeTypeHandle.ToIntPtr` is JIT intrinsic so it will be optimized to just loading a constant in most cases. `RuntimeTypeHandle.Value` does not have that optimization. This change may cause a perf regression.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 9]
```diff
@@ -757,16 +754,13 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
 
 #endif
             }
-            else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-            {
-                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
-            }
 
             GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:758]:
```suggestion
            // Runtime overrides GC_ALLOC_ZEROING_OPTIONAL if the type contains references, so we don't need to worry about that.
            GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
```
I think this is more appropriate place for this comment.

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > > managed value type
> 
> Managed vs. unmanaged types is C# terminology: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/unmanaged-types

Ah. I did not realize they can be applied to an entire value type. That was what I am missing. Thanks.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs [hunk 10]
```diff
@@ -757,16 +754,13 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
             if (pinned)
                 flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
 
-            return Unsafe.As<T[]>(AllocateNewArray(RuntimeTypeHandle.ToIntPtr(typeof(T[]).TypeHandle), length, flags));
+            // Runtime overrides GC_ALLOC_ZEROING_OPTIONAL if the type contains references, so we don't need to worry about that.
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:762]:
```suggestion
```

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (jkotas) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: > LayoutKind.Sequential vs. LayoutKind.Auto

C# marks structs as `LayoutKind.Sequential`. The `LayoutKind.Sequential` applies to the unmanaged view of the struct. It applies to the managed view of the struct only for blittable types where the managed and unmanaged view of struct are the same.

`EmbeddedValueType<string>` and `EmbeddedValueType<object>` types used by this test are not blittable, so the runtime is free to rearrange the fields in the managed view of the struct any way it wants. The current CoreCLR type loader policy is to place all object references first to improve locality for GC heap walking.

The comment "// Adding a simple field to force non-zero object offsets" does not match what is actually going to happen. I agree that it should be reworked.

File: src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs (hunk 8 above)
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:763]: > Oh! I had no idea it was an intrinsic, how is that discoverable?

JIT intrinsics are marked with `[Intrinsic]` attributes.

> Perhaps the should instead revert this line and change the other to use the intrinsic?

Agree.

Comment (jthorborg) [reply src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:763]: Ok, I'll do the change once I get closure on the other discussion

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (jkotas) [reply src/libraries/System.Runtime/tests/System/GCTests.cs:1063]: I think a non-generic struct like this would be sufficient for what you are trying to test here
```
private struct StructType
{
    internal object _;
    internal string _value;
    internal int _;
}
```

File: src/libraries/System.Runtime/tests/System/GCTests.cs [hunk 11]
```diff
@@ -1058,11 +1058,52 @@ private static void AllocateArrayTooLarge()
...
+        [Theory]
+        [InlineData(false), InlineData(true)]
+        private unsafe static void AllocateArrayPinned_ManagedValueType_CanRoundtripThroughPointer(bool uninitialized)
         {
-            GC.AllocateUninitializedArray<string>(100);
-            Assert.Throws<ArgumentException>(() => GC.AllocateUninitializedArray<string>(100, pinned: true));
+            const int length = 100;
+            var rng = new Random(0xAF);
+
+            var array = uninitialized ? GC.AllocateUninitializedArray<EmbeddedValueType<string>>(length, pinned: true) : GC.AllocateArray<EmbeddedValueType<string>>(length, pinned: true);
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1091]:
```suggestion
            EmbeddedValueType<string>[] array = uninitialized ? GC.AllocateUninitializedArray<EmbeddedValueType<string>>(length, pinned: true) : GC.AllocateArray<EmbeddedValueType<string>>(length, pinned: true);
```
Nit: Repo coding conventions

File: src/libraries/System.Runtime/tests/System/GCTests.cs [hunk 12]
```diff
@@ -1058,11 +1058,52 @@ private static void AllocateArrayTooLarge()
...
+            byte* pointer = (byte*)Unsafe.AsPointer(ref array[0]);
+            var size = Unsafe.SizeOf<EmbeddedValueType<string>>();
+
+            for(int i = 0; i < length; ++i)
+            {
+                var idx = rng.Next(length);
+                ref var evt = ref Unsafe.AsRef<EmbeddedValueType<string>>(pointer + size * idx);
+
+                var stringValue = rng.NextSingle().ToString();
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1100]:
```suggestion
            int size = Unsafe.SizeOf<EmbeddedValueType<string>>();

            for(int i = 0; i < length; ++i)
            {
                int idx = rng.Next(length);
                ref EmbeddedValueType<string> evt = ref Unsafe.AsRef<EmbeddedValueType<string>>(pointer + size * idx);

                string stringValue = rng.NextSingle().ToString();
```
Nit: Repo coding conventions. (https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md - We only use var when the type is explicitly named on the right-hand side.)

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs [hunk 13]
```diff
@@ -851,18 +844,12 @@ static T[] AllocateNewUninitializedArray(int length, bool pinned)
...
         {
             GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_NO_FLAGS;
 
             if (pinned)
             {
-                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-                    ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
-
                 flags = GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
             }
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs:855]: Nit: You can replace `EETypePtr.EETypePtrOf<T[]>().RawValue` below with just `MethodTable.Of<T[]>()` while you are on it. It is more terse way to get the MethodTable of specific type in nativeaot.

File: src/libraries/System.Runtime/tests/System/GCTests.cs (hunk 6 above)
Comment (jthorborg) [reply src/libraries/System.Runtime/tests/System/GCTests.cs]: Tried to reach a compromise between all suggestions and added an explanatory comment.

**Comment by MichalPetryka** [comment]
I'd maybe remove the `if` here https://github.com/dotnet/runtime/blob/349970a025a343cd73f5dfba95803afa6187b329/src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs#L740-L743

File: src/libraries/System.Runtime/tests/System/GCTests.cs [hunk 14]
```diff
@@ -1058,11 +1058,59 @@ private static void AllocateArrayTooLarge()
             Assert.Throws<OutOfMemoryException>(() => GC.AllocateUninitializedArray<double>(int.MaxValue, pinned: true));
         }
 
-        [Fact]
-        private static void AllocateArrayRefType()
+        struct EmbeddedValueType<T>
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1061]:
```suggestion
        [StructLayout(LayoutKind.Sequential)]
        struct EmbeddedValueType<T>
```
This should suppress the warning

File: src/libraries/System.Runtime/tests/System/GCTests.cs [hunk 15]
```diff
@@ -1058,11 +1058,59 @@ private static void AllocateArrayTooLarge()
...
+        {
+            // There's a few extra fields here to ensure any reads and writes to Value reasonably are not just offset 0.
+            // This is a low-level smoke check that can give a bit of confidence in pointer arithmetic.
+            // This isn't a guarantee since CoreCLR reorders fields and ignores the implicit sequential consistency of
+            // structs, but it will never hurt the test.
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime/tests/System/GCTests.cs:1066]:
```suggestion
            // The CLR is permitted to reorder fields and ignore the implicit sequential consistency of
            // value types if they contain managed references, but it will never hurt the test.
```

**Comment by jthorborg** [comment]
> I'd maybe remove the `if` here
> 
> https://github.com/dotnet/runtime/blob/349970a025a343cd73f5dfba95803afa6187b329/src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs#L740-L743


Wouldn't that leave dead code just below when `!DEBUG`? Or are you saying remove the whole block instead of just the conditional?

https://github.com/dotnet/runtime/blob/349970a025a343cd73f5dfba95803afa6187b329/src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs#L745-L755

**Comment by jkotas** [comment]
> Wouldn't that leave dead code just below when !DEBUG? Or are you saying remove the whole block instead of just the conditional?

It looks like that this was fine-tuned based on microbenchmarks. If you delete anything here, you will most likely introduce regressions in micro-benchmarks that exercise this API.

I do not think that this is worth touching. You would have to come up with perf numbers that demonstrate that you are not regressing anything that matters.

**Comment by jkotas** [comment]
The wasm build failure is known infrastructure error according to the build analysis.

**Comment by jkotas** [comment]
Thank you!

**Comment by jthorborg** [comment]
Thank you, especially for the handholding while making green contributions. I'm sure it will be smoother the next time I'll have the pleasure.

However trivial this may seem, it will mean a great deal to us.
